src/test/java/com/diffmin/AppTest.java

```diff
@@ -217,5 +266,30 @@ private static void runTests(TestResources sources) throws Exception {
             String patchedProgram = SpoonUtil.displayModifiedModel(patchedCtModel);
             assertEquals(cu.prettyprint(), patchedProgram, "Prev file was not patched correctly");
         }
+
+        // check the root origination of each element
+        Set<String> newRevisionPathStrings =
+                new HashSet<>(Files.readAllLines(sources.newRevisionPaths));
+
+        (new CtScanner() {
+                    @Override
+                    public void scan(CtElement element) {
+                        if (!skipAssertionCheck(element, newRevisionPathStrings)) {
+                            String elementPathString = element.getPath().toString();
+                            if (newRevisionPathStrings.contains(elementPathString)
+                                    || doesElementBelongToModifiedSet(element)) {
+                                assertTrue(
+                                        doesElementBelongToSpecifiedFile(element, NEW_PREFIX),
+                                        "Element should originate from new file but does not");
+                            } else {
+                                assertTrue(
+                                        doesElementBelongToSpecifiedFile(element, PREV_PREFIX),
+                                        "Element should originate from prev file but does not");
+                            }
+                        }
+                        super.scan(element);
+                    }
+                })
+                .scan(patchedCtModel.getRootPackage());
```

Comment:
I still find this unreadable. There is no way for me to tell what this scanner does without reading the entire implementation.  Can you motivate why you use an anonymous class here?

I would put that into a _named_ class at the bottom of the file, and then call it like so.

```suggestion
        new ElementSourceFileChecker(newRevisionPaths).scan(patchedCtModel.getRootPackage());
```

That's still not _great_, as `scan` is a very generic method name, but it's much better just for the fact that the anonymous class is no longer cluttering the core test method.

Reply:
> Can you motivate why you use an anonymous class here?

- This scanner has a very specific use case and I don't think so it will ever be used to facilitate other test methods so I did not bother to extract this into a separate Java class. Even the [Java documentation](https://docs.oracle.com/javase/tutorial/java/javaOO/anonymousclasses.html) says that - "Use them if you need to use a local class only once.".
- I even did this with `InsertionUtil` because my code would never need that logic for any other cases.

Reply:
One more reason I am recalling now is that moving the scanner to another class would require me to change the access modifiers of `PREV_PREFIX` and `NEW_PREFIX` to pacakge-private so I refrained from doing that. There is no specific reason for it except I wanted minimal and only the required changes in the PR.

Reply:
> This scanner has a very specific use case

That's very good, it means that it does one thing.

> and I don't think so it will ever be used to facilitate other test methods so I did not bother to extract this into a separate Java class.

You have several methods in there that are only used once, yet they are methods of their own, so this reasoning doesn't hold for the rest of the test suite. Giving something a name inside of a class or method is not only about code reuse, it is equally about code readability. So just because something is only executed once does not mean that it doesn't deserve to be given a name and a place of its own.

Note that making a named class does not mean that you need to create a new file. You can have as many non-public classes as you'd like in the same file, and in this case a nested class is reasonable to encapsulate.

> Even the Java documentation says that - "Use them if you need to use a local class only once.".

That's the _only_ time you should use them. It doesn't mean that every time you use a class once, you should make it anonymous. And I'd argue that in the vast majority of cases, anonymous classes make for very poor readability.

> I even did this with InsertionUtil because my code would never need that logic for any other cases

I know, I didn't comment on it then because we hadn't really talked too much about readability, but that use of an anonymous class I'd say is the epitome of how not to use it. The code deliberately sets up for the use of an anonymous class, rather than using one where it is appropriate.

> One more reason I am recalling now is that moving the scanner to another class would require me to change the access modifiers of PREV_PREFIX and NEW_PREFIX to package-private so I refrained from doing that.

I don't really see how that's a problem, but you don't have to do that if you create a nested class.

```java
    private static class ElementSourceFileChecker extends CtScanner {
        private final Set<String> newRevisionPathStrings;
        
        ElementSourceFileChecker(Set<String> newRevisionPathStrings) {
            this.newRevisionPathStrings = Set.copyOf(newRevisionPathStrings);
        }

        @Override
        public void scan(CtElement element) {
            if (!skipAssertionCheck(element, newRevisionPathStrings)) {
                String elementPathString = element.getPath().toString();
                if (newRevisionPathStrings.contains(elementPathString)
                        || doesElementBelongToModifiedSet(element)) {
                    assertTrue(
                            doesElementBelongToSpecifiedFile(element, NEW_PREFIX),
                            "Element should originate from new file but does not");
                } else {
                    assertTrue(
                            doesElementBelongToSpecifiedFile(element, PREV_PREFIX),
                            "Element should originate from prev file but does not");
                }
            }
            super.scan(element);
        }

        private static boolean doesElementBelongToSpecifiedFile(
                CtElement element, String filePathPrefix) {
            return element.getPosition().getFile().getName().startsWith(filePathPrefix);
        }

        private static boolean doesElementBelongToModifiedSet(CtElement element) {
            if (element.getRoleInParent() == CtRole.THROWN) {
                return ((CtExecutable<?>) element.getParent())
                        .getThrownTypes().stream()
                        .anyMatch(
                                thrownType ->
                                        doesElementBelongToSpecifiedFile(
                                                thrownType, NEW_PREFIX));
            }
            return false;
        }

        private static boolean isChildOfInsertedPath(
                String elementPathString, Set<String> newRevisionPathStrings) {
            return newRevisionPathStrings.stream()
                    .anyMatch(
                            modifiedPathString ->
                                    !(modifiedPathString.equals(elementPathString))
                                            && elementPathString.startsWith(modifiedPathString));
        }

        private static boolean skipAssertionCheck(
                CtElement element, Set<String> newRevisionPathStrings) {
            if (element == null || element.isImplicit() || !element.getPosition().isValidPosition()) {
                return true;
            }
            String elementPathString = element.getPath().toString();
            return isChildOfInsertedPath(elementPathString, newRevisionPathStrings);
        }

    }
```

Reply:
> I don't really see how that's a problem, but you don't have to do that if you create a nested class.

Yeah, I was creating the same thing but in separate package-private class. Your point here is that when we name the scanner `ElementSourceFileChecker`, the reader would know what the scanner is doing and it will be up to them to read the implementation as well. However, in cases of anonymous class, we compelled the reader to read the implementation. I just want to confirm.

Reply:
Kind of. Here are the problems I see with the anonymous class.

1. It has no name, and so there's absolutely no indication of what it does other than its implementation.
2. I have to read the entire implementation of the anonymous class before I even see which method is called on it, and with what arguments.
3. The anonymous class being defined inside of the test method (which is already too large, doing too many things) makes said test a lot harder to just glance over.

It would be even more readable if wrapped `scan` with some more appropriately named method (`ElementSourceFileChecker` doesn't really convey the entire message), perhaps `assertElementsHaveExpectedSourceFiles()`.

Reply:
> It would be even more readable if wrapped scan with some more appropriately named method (ElementSourceFileChecker doesn't really convey the entire message), perhaps assertElementsHaveExpectedSourceFiles().

Sounds good. I will keep this in mind while refactoring.